presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java

```diff
@@ -0,0 +1,332 @@
+                .toString();
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(values) + sizeOf(nulls) + sizeOf(outputPositions);
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        boolean nullsAllowed = filter == null || filter.testNull();
+        boolean recordNulls = nullsAllowed && presentStream != null;
+        if (outputRequired) {
+            ensureValuesCapacity(positionCount, recordNulls);
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
+        }
+        else {
+            outputPositions = positions;
+        }
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
+        allNulls = false;
+        if (dataStream == null && presentStream != null) {
+            // all nulls
+            presentStream.skip(positions[positionCount - 1]);
+            streamPosition = positions[positionCount - 1] + 1;
+
+            if (nullsAllowed) {
+                outputPositionCount = positionCount;
+                allNulls = true;
+            }
+            else {
+                outputPositionCount = 0;
+            }
+        }
+        else if (filter == null) {
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    nulls[i] = true;
+                }
+                else {
+                    values[i] = (byte) (dataStream.nextBit() ? 1 : 0);
+                    if (recordNulls) {
+                        nulls[i] = false;
+                    }
+                }
+                streamPosition++;
+            }
+            outputPositionCount = positionCount;
+        }
+        else {
+            outputPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    if (nullsAllowed) {
+                        if (outputRequired) {
```

Comment:
Do we wanna check `recordNulls`? There are too many flag combinations. Having helpers would be easier to understand.

Reply:
- `isNull == true` must be recorded all the time; only `isNull == false` recording may be unnecessary
- would it help to rename `recordNulls` to `recordNotNull`?